src/tracing.rs

```diff
@@ -68,5 +43,371 @@ pub fn setup(user_filters: &str) -> anyhow::Result<()> {
+        fn from(events: SpanEvents) -> Self {
+            events.to_fmt_span().into()
+        }
+    }
+
+    #[derive(Copy, Clone, Debug, Deserialize, Serialize)]
+    #[serde(rename_all = "lowercase")]
+    pub enum SpanEvent {
+        New,
+        Enter,
+        Exit,
+        Close,
+    }
+
+    impl SpanEvent {
+        pub fn from_fmt_span(events: fmt::format::FmtSpan) -> Vec<Self> {
+            const EVENTS: &[SpanEvent] = &[
+                SpanEvent::New,
+                SpanEvent::Enter,
+                SpanEvent::Exit,
+                SpanEvent::Close,
+            ];
+
+            EVENTS
+                .iter()
+                .copied()
+                .filter(|event| events.clone() & event.to_fmt_span() == event.to_fmt_span())
+                .collect()
+        }
+
+        pub fn to_fmt_span(self) -> fmt::format::FmtSpan {
+            match self {
+                SpanEvent::New => fmt::format::FmtSpan::NEW,
+                SpanEvent::Enter => fmt::format::FmtSpan::ENTER,
+                SpanEvent::Exit => fmt::format::FmtSpan::EXIT,
+                SpanEvent::Close => fmt::format::FmtSpan::CLOSE,
+            }
+        }
+    }
+
+    #[derive(Copy, Clone, Debug, Deserialize, Serialize)]
+    #[serde(rename_all = "lowercase")]
+    pub enum NoneTag {
+        None,
+    }
+
+    #[derive(Copy, Clone, Debug, Deserialize, Serialize)]
+    #[serde(untagged)]
+    pub enum Color {
+        Auto(AutoTag),
+        Bool(bool),
+    }
+
+    impl From<super::Color> for Color {
+        fn from(color: super::Color) -> Self {
+            match color {
+                super::Color::Auto => Self::Auto(AutoTag::Auto),
+                super::Color::Enable => Self::Bool(true),
+                super::Color::Disable => Self::Bool(false),
+            }
+        }
+    }
+
+    impl From<Color> for super::Color {
+        fn from(color: Color) -> Self {
+            match color {
+                Color::Auto(_) => Self::Auto,
+                Color::Bool(true) => Self::Enable,
+                Color::Bool(false) => Self::Disable,
+            }
+        }
+    }
+
+    #[derive(Copy, Clone, Debug, Deserialize, Serialize)]
+    #[serde(rename_all = "lowercase")]
+    pub enum AutoTag {
+        Auto,
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize)]
+pub struct LoggerConfigDiff {
+    #[serde(flatten)]
+    default: DefaultLoggerConfigDiff,
+}
+
+#[derive(Clone, Debug, Default, Deserialize)]
+pub struct DefaultLoggerConfigDiff {
+    log_level: Option<String>,
+    span_events: Option<SpanEvents>,
+    color: Option<Color>,
+}
+
+const DEFAULT_LOG_LEVEL: log::LevelFilter = log::LevelFilter::Info;
+
+const DEFAULT_FILTERS: &[(&str, log::LevelFilter)] = &[
+    ("hyper", log::LevelFilter::Info),
+    ("h2", log::LevelFilter::Error),
+    ("tower", log::LevelFilter::Warn),
+    ("rustls", log::LevelFilter::Info),
+    ("wal", log::LevelFilter::Warn),
+    ("raft", log::LevelFilter::Warn),
+];
```

Comment:
We already use this, though. It's done so that these modules, that have very verbose logs, would *not* be enabled with the "global" `debug` level-filter, unless *explicitly* enabled `debug,hyper=debug`.

Reply:
Cool stuff. Thanks for the explanation!

Reply:
:raised_hands:

Reply:
Had to come up with this hack, because `tracing_subscriber::filter::EnvFilter` does not provide enough methods to configure it properly. :/